File: LJI/스터디-C++/main_1041_주사위.cpp
#include <iostream>
#include <vector>

using namespace std;

int N;

// N=1일 때는 예외 적용

// 연결된 3개의 면의 최소값 *4
// 연결된 2개 면의 최소값 (N-1)*4+(N-2)*4
// 가장 작은 한개의 면 (N-1)*N-2*4 + (N-2)*(N-2)

// 5-i가 반대면
main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    cin >> N;
    vector<int> Dice(5);

    int Min1 = 1000;
    int Min2 = 1000;
    int Min3 = 1000;
    int Max1 = 0;
    long long Sum = 0;
    for (int i = 0; i < 6; i++)
    {
        cin >> Dice[i];
        Min1 = min(Min1, Dice[i]);
        Max1 = max(Max1, Dice[i]);
        Sum += Dice[i];
    }

    if (N == 1)
    {
        cout << Sum - Max1;
        return 0;
    }

    for (int i = 0; i < 6; i++)
    {
        for (int j = i + 1; j < 6; j++)
        {
            // 반대면 제외
            if (5 - j != i)
            {
                Min2 = min(Min2, Dice[i] + Dice[j]);
            }

            // 3면 배치
            for (int k = j + 1; k < 6; k++)
            {
                if (5 - k == i || 5 - k == j || 5 - j == i)
                    continue;

                Min3 = min(Min3, Dice[i] + Dice[j] + Dice[k]);
            }
        }
    }

    Sum = 0;
    // 연결된 3개의 면의 최소값 *4
    Sum += 4 * Min3;
    // 연결된 2개 면의 최소값 (N-1)*4+(N-2)*4
    Sum += (long long)((N - 1) * 4 + (N - 2) * 4) * Min2;
    // 가장 작은 한개의 면 (N-1)*N-2*4 + (N-2)*(N-2)
    Sum += ((long long)(N - 1) * (N - 2) * 4 + (long long)(N - 2) * (N - 2)) * Min1;

    cout << Sum;
}

File: JiHwanYoon/Main_1041_주사위.java


import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main_1041_주사위 {
	public static void main(String[] args) throws IOException {
		// 입력
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		String[] s = br.readLine().split(" ");
		int[] nums = new int[6];
		for (int i = 0; i < 6; i++) {
			nums[i] = Integer.parseInt(s[i]);
		}
		// 그리디 알고리즘 활용
		// N*N*N 크기의 정육면체를 만들면, 한쪽 면만 보이는 주사위, 두쪽 면만 보이는 주사위, 세쪽 면만 보이는 주사위가 있다.
		// 한쪽 면만 보이는 주사위 개수는 총 5*(N-2)*(N-2)(5개의 면 중앙 (N-2)*(N-2) 정사각형) + 4*(N-2)(아래쪽 면과 접하는 모서리 부분 (N-2)개 * 4)
		// 두쪽 면만 보이는 주사위 개수는 총 8*N-12(아래쪽 면과 접하지 않는 모서리 부분 (N-2)개 * 8 + 아래쪽 면과 접하는 꼭짓점 부분 4개)
		// 세쪽 면만 보이는 주사위 개수는 총 4개(아래쪽 면과 접하는 꼭짓점 부분 4개)
		long min1 = Integer.MAX_VALUE;
		long min2 = Integer.MAX_VALUE;
		long min3 = Integer.MAX_VALUE;
		// 한쪽 면만 보이는 주사위의 경우 가장 작은 값만 보이도록 하면 된다.
		for (int i = 0; i < 6; i++) {
			min1 = Math.min(min1, nums[i]);
		}
		// 두쪽 면 이상 보이는 주사위의 경우 주사위 방향에 주의해 최솟값을 구한다.
		min2 = Math.min(min2, nums[0]+nums[1]+nums[2]);
		min2 = Math.min(min2, nums[0]+nums[1]+nums[3]);
		min2 = Math.min(min2, nums[0]+nums[3]+nums[4]);
		min2 = Math.min(min2, nums[0]+nums[2]+nums[4]);
		min2 = Math.min(min2, nums[5]+nums[1]+nums[2]);
		min2 = Math.min(min2, nums[5]+nums[1]+nums[3]);
		min2 = Math.min(min2, nums[5]+nums[3]+nums[4]);
		min2 = Math.min(min2, nums[5]+nums[2]+nums[4]);
		min3 = Math.min(min3, nums[0]+nums[1]);
		min3 = Math.min(min3, nums[0]+nums[2]);
		min3 = Math.min(min3, nums[0]+nums[3]);
		min3 = Math.min(min3, nums[0]+nums[4]);
		min3 = Math.min(min3, nums[1]+nums[2]);
		min3 = Math.min(min3, nums[1]+nums[3]);
		min3 = Math.min(min3, nums[1]+nums[5]);
		min3 = Math.min(min3, nums[2]+nums[4]);
		min3 = Math.min(min3, nums[2]+nums[5]);
		min3 = Math.min(min3, nums[3]+nums[4]);
		min3 = Math.min(min3, nums[3]+nums[5]);
		min3 = Math.min(min3, nums[4]+nums[5]);
		long sum = 0L;
		if (N == 1) {
			int max = Integer.MIN_VALUE;
			for (int i = 0; i < 6; i++) {
				max = Math.max(max, nums[i]);
				sum += nums[i];
			}
			sum -= max;
		} else {
			sum = min1*5*(N-2)*(N-2)+min1*4*(N-2)+min3*(8*N-12)+min2*4;
		}
		System.out.println(sum);
	}

}
